File: Shared/AgeHelper.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Class AgeHelper

    Private m_IsEnabled As Boolean
    Private m_Frequency As Integer
    Private m_Maximum As Integer

    Const AGE_KEY_DEFAULT As Integer = 0
    Const MAX_AGE_CLASSES As Integer = 1000

    ''' <summary>
    ''' Constructor
    ''' </summary>
    ''' <param name="isEnabled"></param>
    ''' <param name="frequency"></param>
    ''' <param name="maximum"></param>
    ''' <remarks></remarks>
    Public Sub New(
        ByVal isEnabled As Boolean,
        ByVal frequency As Integer,
        ByVal maximum As Integer)

        If (isEnabled) Then

            If (frequency <= 0) Then

                ExceptionUtils.ThrowArgumentException(
                    "The age reporting frequency must be greater than zero.")

            End If

            If (maximum < frequency) Then

                ExceptionUtils.ThrowArgumentException(
                    "The maximum age cannot be less than the age reporting frequency.")

            End If

        End If

        Me.m_IsEnabled = isEnabled
        Me.m_Frequency = frequency
        Me.m_Maximum = maximum

        If ((Me.m_Maximum / Me.m_Frequency) > MAX_AGE_CLASSES) Then
            Me.m_Maximum = (Me.m_Frequency * MAX_AGE_CLASSES)
        End If

    End Sub

    ''' <summary>
    ''' Gets a key for the specified age
    ''' </summary>
    ''' <param name="age"></param>
    ''' <returns></returns>
    ''' <remarks>
    ''' If enabled then the key is the minimum age, otherwise it is the default.
    ''' </remarks>
    Public Function GetKey(ByVal age As Integer) As Integer

        If (Not Me.m_IsEnabled) Then
            Return AGE_KEY_DEFAULT
        Else
            Return Me.GetAgeMinimum(age).Value
        End If

    End Function

    ''' <summary>
    ''' Gets the minimum age
    ''' </summary>
    ''' <param name="age"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetAgeMinimum(ByVal age As Integer) As Nullable(Of Integer)

        If (Not Me.m_IsEnabled) Then
            Return Nothing
        Else

            If (age > Me.m_Maximum) Then
                Return ((Me.m_Maximum \ Me.m_Frequency) * Me.m_Frequency)
            Else
                Return ((age \ Me.m_Frequency) * Me.m_Frequency)
            End If

        End If

    End Function

    ''' <summary>
    ''' Gets the reporting age maximum for the specified age
    ''' </summary>
    ''' <param name="age"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetAgeMaximum(ByVal age As Integer) As Nullable(Of Integer)

        If (Not Me.m_IsEnabled) Then
            Return Nothing
        Else

            Dim a As Nullable(Of Integer) = Me.GetAgeMinimum(age) + (Me.m_Frequency - 1)

            If (a.Value >= Me.m_Maximum) Then
                a = Nothing
            End If

            Return a

        End If

    End Function

    ''' <summary>
    ''' Gets an enumeration of age descriptors for the current configuration
    ''' </summary>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetAges() As IEnumerable(Of AgeDescriptor)

        Dim v As Integer = 0
        Dim lst As New List(Of AgeDescriptor)

        While (v <= Me.m_Maximum)

            Dim min As Integer = Me.GetAgeMinimum(v).Value
            Dim max As Nullable(Of Integer) = Me.GetAgeMaximum(v)

            lst.Add(New AgeDescriptor(min, max))
            v += Me.m_Frequency

        End While

        Debug.Assert(lst.Count > 0)
        Debug.Assert(Not lst(lst.Count - 1).MaximumAge.HasValue)

        Return lst

    End Function

End Class

''' <summary>
''' Age Descriptor
''' </summary>
''' <remarks></remarks>
Class AgeDescriptor

    Private m_MinimumAge As Integer
    Private m_MaximumAge As Nullable(Of Integer)

    Public Sub New(
        ByVal minimumAge As Integer,
        ByVal maximumAge As Nullable(Of Integer))

        Me.m_MinimumAge = minimumAge
        Me.m_MaximumAge = maximumAge

    End Sub

    Public Property MinimumAge As Integer
        Get
            Return Me.m_MinimumAge
        End Get
        Set(value As Integer)
            Me.m_MinimumAge = value
        End Set
    End Property

    Public Property MaximumAge As Nullable(Of Integer)
        Get
            Return Me.m_MaximumAge
        End Get
        Set(value As Nullable(Of Integer))
            Me.m_MaximumAge = value
        End Set
    End Property

End Class

File: src/Shared/AgeHelper.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System.Diagnostics;
using System.Collections.Generic;

namespace SyncroSim.STSim
{
    internal class AgeHelper
    {
        private bool m_IsEnabled;
        private int m_Frequency;
        private int m_Maximum;
        private const int AGE_KEY_DEFAULT = 0;
        private const int MAX_AGE_CLASSES = 1000;

        public AgeHelper(bool isEnabled, int frequency, int maximum)
        {
            if (isEnabled)
            {
                if (frequency <= 0)
                {
                    ExceptionUtils.ThrowArgumentException("The age reporting frequency must be greater than zero.");
                }

                if (maximum < frequency)
                {
                    ExceptionUtils.ThrowArgumentException("The maximum age cannot be less than the age reporting frequency.");
                }
            }

            this.m_IsEnabled = isEnabled;
            this.m_Frequency = frequency;
            this.m_Maximum = maximum;

            if ((this.m_Maximum / (double)this.m_Frequency) > MAX_AGE_CLASSES)
            {
                this.m_Maximum = (this.m_Frequency * MAX_AGE_CLASSES);
            }
        }

        /// <summary>
        /// Gets a key for the specified age
        /// </summary>
        /// <param name="age"></param>
        /// <returns></returns>
        /// <remarks>
        /// If enabled then the key is the minimum age, otherwise it is the default.
        /// </remarks>
        public int GetKey(int age)
        {
            if (!this.m_IsEnabled)
            {
                return AGE_KEY_DEFAULT;
            }
            else
            {
                return this.GetAgeMinimum(age).Value;
            }
        }

        /// <summary>
        /// Gets the minimum age
        /// </summary>
        /// <param name="age"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public int? GetAgeMinimum(int age)
        {
            if (!this.m_IsEnabled)
            {
                return null;
            }
            else
            {
                if (age > this.m_Maximum)
                {
                    return ((this.m_Maximum / this.m_Frequency) * this.m_Frequency);
                }
                else
                {
                    return ((age / this.m_Frequency) * this.m_Frequency);
                }
            }
        }

        /// <summary>
        /// Gets the reporting age maximum for the specified age
        /// </summary>
        /// <param name="age"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public int? GetAgeMaximum(int age)
        {
            if (!this.m_IsEnabled)
            {
                return null;
            }
            else
            {
                int? a = this.GetAgeMinimum(age) + (this.m_Frequency - 1);

                if (a.Value >= this.m_Maximum)
                {
                    a = null;
                }

                return a;
            }
        }

        /// <summary>
        /// Gets an enumeration of age descriptors for the current configuration
        /// </summary>
        /// <returns></returns>
        /// <remarks></remarks>
        public IEnumerable<AgeDescriptor> GetAges()
        {
            int v = 0;
            List<AgeDescriptor> lst = new List<AgeDescriptor>();

            while (v <= this.m_Maximum)
            {
                int min = this.GetAgeMinimum(v).Value;
                int? max = this.GetAgeMaximum(v);

                lst.Add(new AgeDescriptor(min, max));
                v += this.m_Frequency;
            }

            Debug.Assert(lst.Count > 0);
            Debug.Assert(!(lst[lst.Count - 1].MaximumAge.HasValue));

            return lst;
        }
    }

    /// <summary>
    /// Age Descriptor
    /// </summary>
    /// <remarks></remarks>
    internal class AgeDescriptor
    {
        private int m_MinimumAge;
        private int? m_MaximumAge;

        public AgeDescriptor(int minimumAge, int? maximumAge)
        {
            this.m_MinimumAge = minimumAge;
            this.m_MaximumAge = maximumAge;
        }

        public int MinimumAge
        {
            get
            {
                return this.m_MinimumAge;
            }
            set
            {
                this.m_MinimumAge = value;
            }
        }

        public int? MaximumAge
        {
            get
            {
                return this.m_MaximumAge;
            }
            set
            {
                this.m_MaximumAge = value;
            }
        }
    }
}
